src/webgpu/api/operation/rendering/3d_texture_slices.spec.ts

```diff
@@ -0,0 +1,363 @@
+    attachmentCount: number,
+    buffer: GPUBuffer
+  ) {
+    const { bufferSize, bufferOffset } = this.getBufferSizeAndOffset(
+      attachmentWidth,
+      attachmentHeight,
+      attachmentCount
+    );
+    const expectedData = new Uint8Array(bufferSize);
+    for (let i = 0; i < attachmentCount; i++) {
+      for (let j = 0; j < attachmentHeight; j++) {
+        for (let k = 0; k < attachmentWidth; k++) {
+          expectedData[i * bufferOffset + j * 256 + k * 4] = k < j ? 0x00 : 0xff;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 1] = k < j ? 0xff : 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 2] = 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 3] = 0xff;
+        }
+      }
+    }
+
+    this.expectGPUBufferValuesEqual(buffer, expectedData);
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('one_color_attachment,mip_levels')
+  .desc(
+    `
+  Render to a 3d texture slice with mip levels.
+  `
+  )
+  .params(u => u.combine('mipLevel', [0, 1, 2]).combine('depthSlice', [0, 1]))
+  .fn(t => {
+    const { mipLevel, depthSlice } = t.params;
+
+    const texture = t.create3DTexture({
+      width: kSize << mipLevel,
+      height: kSize << mipLevel,
+      depthOrArrayLayers: 2 << mipLevel,
+      mipLevelCount: mipLevel + 1,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+    });
+
+    const { bufferSize } = t.getBufferSizeAndOffset(kSize, kSize, 1);
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+    const pipeline = t.createRenderPipeline();
+
+    const colorAttachment = t.getColorAttachment(texture, {
+      baseMipLevel: mipLevel,
+      mipLevelCount: 1,
+    });
+    colorAttachment.depthSlice = depthSlice;
+
+    const encoder = t.createEncoder('non-pass');
+    const pass = encoder.encoder.beginRenderPass({ colorAttachments: [colorAttachment] });
+    pass.setPipeline(pipeline);
+    pass.draw(3);
+    pass.end();
+    encoder.encoder.copyTextureToBuffer(
+      { texture, mipLevel, origin: { x: 0, y: 0, z: depthSlice } },
+      { buffer, bytesPerRow: 256 },
+      { width: kSize, height: kSize, depthOrArrayLayers: 1 }
+    );
+    t.device.queue.submit([encoder.finish()]);
+
+    t.checkAttachmentResult(kSize, kSize, 1, buffer);
+  });
+
+g.test('multiple_color_attachments,same_mip_level')
+  .desc(
+    `
+  Render to the different slices of 3d texture in multiple color attachments.
+  - Same 3d texture with different slices at same mip level
+  - Different 3d textures with same slice at same mip level
+  `
+  )
+  .params(u =>
+    u
+      .combine('sameTexture', [true, false])
+      .beginSubcases()
+      .combine('samePass', [true, false])
+      .combine('mipLevel', [0, 1])
+  )
+  .fn(t => {
+    const { sameTexture, samePass, mipLevel } = t.params;
+
+    const format = 'rgba8unorm' as GPUTextureFormat;
```

Comment:
Always use `as const` for string enums (it _is_ safe)
```suggestion
    const format = 'rgba8unorm' as const;
```